apps/api/src/notifications/notifications.service.ts

```diff
@@ -245,7 +228,148 @@ export class MarketingNotificationsService {
+    }
+    // If registered and no existing consent -> user is already unsubscribed
+    if (registered && !registered.newsletter) return { message: 'Unsubscribed' }
+
+    // Check if un-registered consent already exists for this email
+    let unregisteredConsent: UnregisteredNotificationConsent | null
+    try {
+      unregisteredConsent = await this.prisma.unregisteredNotificationConsent.findFirst({
+        where: { email: data.email },
+      })
+    } catch (e) {
+      return new BadRequestException('Failed to get email data')
+    }
+    // If unregistered consent exists -> user is already subscribed
+    if (unregisteredConsent && !unregisteredConsent.consent) return { message: 'Unsubscribed' }
+
+    if (!registered && !unregisteredConsent) throw new BadRequestException('Invalid email')
+
+    //If the email wants a general unsubscribe from all notifications
+    // Add to unsubscribed list for all notifications
+    if (!data.campaignId)
+      try {
+        await this.marketingNotificationsProvider.addToUnsubscribed({ emails: [data.email] })
+      } catch (e) {
+        Logger.error('Failed to unsubscribe email', e)
+        throw new BadRequestException('Failed to unsubscribe email')
+      }
+    // else remove it only from the particular campaign list
+    else if (data.campaignId)
+      try {
+        //get campaign notification list
+        const campaign = await this.prisma.campaign.findFirst({
+          where: { id: data.campaignId },
+          select: { notificationLists: true },
+        })
+
+        if (!campaign) return new NotFoundException('Campaign not found')
+
+        // if such a list exists
+        if (campaign?.notificationLists?.length) {
+          // find the email id in the marketing platform
+          const contactsInfo = await this.marketingNotificationsProvider.getContactsInfo({
+            emails: [data.email],
+          })
+          // remove from the campaign notification list
+          Object.keys(contactsInfo) &&
+            (await this.marketingNotificationsProvider.removeContactsFromList({
+              list_id: campaign.notificationLists[0].id,
+              contact_ids: [contactsInfo[data.email]?.contact.id],
+            }))
+        }
+      } catch (e) {
+        Logger.error('Failed to unsubscribe email', e)
+        throw new BadRequestException('Failed to unsubscribe email')
+      }
+
+    // update consent status
+    try {
+      if (!data.campaignId && registered)
+        await this.personService.update(registered.id, { newsletter: false })
+      if (!data.campaignId && unregisteredConsent)
+        await this.personService.updateUnregisteredNotificationConsent(data.email, false)
+    } catch (e) {
+      Logger.error('Failed to update email consent', e)
+      throw new BadRequestException('Failed to update email consent')
+    }
+
+    return { message: 'Success' }
+  }
+
+  async unsubscribe(data: UnsubscribeDto, user: KeycloakTokenParsed) {
```

Comment:
could you check if possible to extract common parts between unsubscribePublic() and unsubscribe() ?

Reply:
Done